### PR TITLE
Data lifecycle PUT/GET/DELETE APIs

### PR DESCRIPTION
**Pull request by nguptaopensds**
**What this PR does / why we need it**: This PR is enhancement to the APIs for lifecycle

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
PUT and GET refactoring and Delete API is fixed

**Special notes for your reviewer**:
PUT:
```
http://127.0.0.1:8089/v1/s3/test/?lifecycle
<LifecycleConfiguration>
  <Rule>
    <ID>rid1</ID>
    <Filter>
       <Prefix>documents/</Prefix>
    </Filter>
    <Status>Enabled</Status>
    <Transition>
      <Days>30</Days>
      <StorageClass>STANDARD</StorageClass>
      <Backend>AWS S3</Backend>
    </Transition>
  </Rule>
  <Rule>
    <ID>rid2</ID>
    <Filter>
       <Prefix>logs/</Prefix>
    </Filter>
    <Status>Enabled</Status>
    <Expiration>
      <Days>365</Days>
    </Expiration>
    <AbortIncompleteMultipartUpload>
    	<DaysAfterInitiation>35</DaysAfterInitiation>
    </AbortIncompleteMultipartUpload>
  </Rule>
</LifecycleConfiguration>
```
GET:
```
http://127.0.0.1:8089/v1/s3/test/?lifecycle
```
DELETE:
```
Can specify the rule id which you wish to delete
http://127.0.0.1:8089/v1/s3/test/?lifecycle&ruleID=rid1&ruleID=rid2
```
**Release note**:

```release-note
```

### CODE REVIEW

File: api/pkg/s3/bucketlifecycleget.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+		xmlRule.Status = lcRule.Status
+		xmlRule.ID = lcRule.ID
+		xmlRule.Filter = converts3FilterToRuleFilter(lcRule.Filter)
+		xmlRule.AbortIncompleteMultipartUpload = converts3UploadToRuleUpload(lcRule.AbortIncompleteMultipartUpload)
+		xmlRule.Transition = make([]model.Transition, 0)
+		for _, transition := range lcRule.Actions {
+			xmlTransition := model.Transition{}
+			xmlTransition.Days = transition.Days
+			//xmlTransition.StorageClass = transition.Tier
```
Comment (wisererik) [inline api/pkg/s3/bucketlifecycleget.go]: remove it

File: api/pkg/s3/bucketlifecycleget.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+			xmlExpiration := model.Expiration{}
+			xmlExpiration.Days = expiration.Days
+			xmlRule.Expiration = append(xmlRule.Expiration, xmlExpiration)
+		}
+		// append each xml rule to xml array
+		getLifecycleConf.Rule = append(getLifecycleConf.Rule, xmlRule)
+	}
+
+	// marshall the array back to xml
+	//xmlByteArr, _ := xml.Marshal(getLifecycleConf.Rule)
```
Comment (wisererik) [inline api/pkg/s3/bucketlifecycleget.go]: same here

File: api/pkg/s3/bucketlifecycleput.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+					s3Transition.Backend = transition.Backend
+
+					//Assigning the storage class of the object to s3 struct
+					//tier, err := s.class2tier(transition.StorageClass)
+					//if err != nil{
+					//	response.WriteError(http.StatusBadRequest, err)
+					//	return
+					//}
+					//s3Transition.Tier = tier
+
```
Comment (wisererik) [inline api/pkg/s3/bucketlifecycleput.go]: same here

File: s3/pkg/db/drivers/mongo/bucketlifecycledelete.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+	//Check if the connector exist or not
+	ss := ad.s.Copy()
+	defer ss.Close()
+
+	//Delete it from database
+	c := ss.DB(DataBaseName).C(BucketMD)
+	log.Logf("bucketName is %v:", in.Bucket)
+	err := c.Update(bson.M{"name": in.Bucket}, bson.M{"$pull": bson.M{"lifecycleconfiguration": bson.M{"id": in.RuleID}}})
+	if err != nil {
+		log.Logf("Delete lifecycle for bucket : %s and lifecycleruleId : %s failed,err:%v.\n", in.Bucket, in.RuleID, err)
```
Comment (sfzeng) [inline s3/pkg/db/drivers/mongo/bucketlifecycledelete.go]: The first letter in error message need to be lowercase.

File: api/pkg/s3/bucketlifecycledelete.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+	"golang.org/x/net/context"
+	"net/http"
+)
+
+func (s *APIService) BucketLifecycleDelete(request *restful.Request, response *restful.Response) {
+	if !policy.Authorize(request, response, "bucket:delete") {
+		return
+	}
+	bucketName := request.PathParameter("bucketName")
+	ruleID := request.Request.URL.Query()["ruleID"]
```
Comment (sfzeng) [inline api/pkg/s3/bucketlifecycledelete.go]: What is the format of ruleID, and what if the input ruleID is with invalid format.

File: api/pkg/s3/bucketlifecycledelete.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+	ctx := context.Background()
+	log.Logf("Received request for bucket lifecycle delete for bucket: %s and the ruleID: %s", bucketName, ruleID)
+	bucket, _ := s.s3Client.GetBucket(ctx, &s3.Bucket{Name: bucketName})
+	for _, lcRule := range bucket.LifecycleConfiguration {
+		for _, id := range ruleID {
+			if lcRule.ID == id {
```
Comment (sfzeng) [inline api/pkg/s3/bucketlifecycledelete.go]: If user input an non-exist ruleID, we need to return some error to let user know that. But with the current implementation, nothing will be returned.

File: api/pkg/s3/bucketlifecycleput.go (hunk 3 above)
Comment (sfzeng) [reply api/pkg/s3/bucketlifecycleput.go]: Storage class is lost here, we need to translate storage class to internal tier and put the tier to s3Tansition.Tier.

File: api/pkg/s3/bucketlifecycleput.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+		if err != nil {
+			response.WriteError(http.StatusInternalServerError, err)
+			return
+		} else {
+			s3RulePtrArr := make([]*s3.LifecycleRule, 0)
+			for _, rule := range createLifecycleConf.Rule {
+				s3Rule := s3.LifecycleRule{}
+
+				// assign the fields
+				s3Rule.ID = rule.ID //Assigning the rule ID
```
Comment (sfzeng) [inline api/pkg/s3/bucketlifecycleput.go]: Duplication check of rule ID is necessary.

File: api/pkg/s3/bucketlifecycleput.go [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+					s3Expiration := s3.Action{Name: "Expiration"}
+					s3Expiration.Days = expiration.Days
+					s3ActionArr = append(s3ActionArr, &s3Expiration)
+				}
+				//Assigning the Expiration action to s3 struct expiration
+				s3Rule.Actions = s3ActionArr
+
+				s3Rule.AbortIncompleteMultipartUpload = convertRuleUploadToS3Upload(rule.AbortIncompleteMultipartUpload)
+				// add to the s3 array
+				s3RulePtrArr = append(s3RulePtrArr, &s3Rule)
```
Comment (sfzeng) [inline api/pkg/s3/bucketlifecycleput.go]: The validation check of rule is necessary, for example, if only expiration action exist, then expiration days should be more than one, if expiration and transition both exist, the expiration days should be more than days of transition for 30 days.

File: api/pkg/s3/bucketlifecycleput.go [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+			}
+			// assign lifecycle rules to s3 bucket
+			bucket.LifecycleConfiguration = s3RulePtrArr
+			log.Logf("final bucket metadata is %v\n", bucket)
+		}
+	}
+
+	// Create bucket with bucket name will check if the bucket exists or not, if it exists
+	// it will internally call UpdateBucket function
+	res, err := s.s3Client.CreateBucket(ctx, bucket)
```
Comment (sfzeng) [inline api/pkg/s3/bucketlifecycleput.go]: Need to update bucket, instead of create bucket again.

File: api/pkg/s3/bucketlifecycleput.go [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+		return
+	}
+	log.Log("Create bucket lifecycle successfully.")
+	response.WriteEntity(res)
+
+}
+
+func convertRuleFilterToS3Filter(filter model.Filter) *s3.LifecycleFilter {
+	retFilter := s3.LifecycleFilter{}
+	retFilter.Prefix = filter.Prefix
```
Comment (sfzeng) [inline api/pkg/s3/bucketlifecycleput.go]: Need to check if filter is nil, otherwise there may panic.

File: api/pkg/s3/bucketlifecycleput.go [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+	return &retFilter
+}
+
+func convertRuleUploadToS3Upload(upload model.AbortIncompleteMultipartUpload) *s3.AbortMultipartUpload {
+	retUpload := s3.AbortMultipartUpload{}
+	retUpload.DaysAfterInitiation = upload.DaysAfterInitiation
```
Comment (sfzeng) [inline api/pkg/s3/bucketlifecycleput.go]: Check nil too.

File: api/pkg/s3/bucketlifecycleget.go [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+	for _, lcRule := range bucket.LifecycleConfiguration {
+		xmlRule := model.Rule{}
+
+		xmlRule.Status = lcRule.Status
+		xmlRule.ID = lcRule.ID
+		xmlRule.Filter = converts3FilterToRuleFilter(lcRule.Filter)
+		xmlRule.AbortIncompleteMultipartUpload = converts3UploadToRuleUpload(lcRule.AbortIncompleteMultipartUpload)
+		xmlRule.Transition = make([]model.Transition, 0)
+		for _, transition := range lcRule.Actions {
+			xmlTransition := model.Transition{}
```
Comment (sfzeng) [inline api/pkg/s3/bucketlifecycleget.go]: Need to check if the rule's name is transition, only transition rule can be add to xmlRule.Transition.

File: api/pkg/s3/bucketlifecycleget.go [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+			xmlTransition.Days = transition.Days
+			//xmlTransition.StorageClass = transition.Tier
+			xmlTransition.Backend = transition.Backend
+			xmlRule.Transition = append(xmlRule.Transition, xmlTransition)
+		}
+		for _, expiration := range lcRule.Actions {
```
Comment (sfzeng) [inline api/pkg/s3/bucketlifecycleget.go]: Need to check if the rule's name is expiration, only expiration rule can be add to xmlRule.Expiration..

File: api/pkg/s3/bucketlifecycleget.go (hunk 1 above)
Comment (nguptaopensds) [reply api/pkg/s3/bucketlifecycleget.go]: Resolved

File: api/pkg/s3/bucketlifecycleget.go (hunk 2 above)
Comment (nguptaopensds) [reply api/pkg/s3/bucketlifecycleget.go]: Resolved

File: api/pkg/s3/bucketlifecycleput.go (hunk 3 above)
Comment (nguptaopensds) [reply api/pkg/s3/bucketlifecycleput.go]: Resolved

File: api/pkg/s3/bucketlifecycleput.go (hunk 7 above)
Comment (nguptaopensds) [reply api/pkg/s3/bucketlifecycleput.go]: Done , please see in the latest commit

File: api/pkg/s3/bucketlifecycleput.go (hunk 8 above)
Comment (nguptaopensds) [reply api/pkg/s3/bucketlifecycleput.go]: Done in the latest commit

File: api/pkg/s3/bucketlifecycleput.go (hunk 9 above)
Comment (nguptaopensds) [reply api/pkg/s3/bucketlifecycleput.go]: Done in the latest commit

File: api/pkg/s3/bucketlifecycleput.go (hunk 10 above)
Comment (nguptaopensds) [reply api/pkg/s3/bucketlifecycleput.go]: Done in the latest commit

File: api/pkg/s3/bucketlifecycleput.go (hunk 11 above)
Comment (nguptaopensds) [reply api/pkg/s3/bucketlifecycleput.go]: Resolved in latest commit

File: s3/pkg/db/drivers/mongo/bucketlifecycledelete.go (hunk 4 above)
Comment (nguptaopensds) [reply s3/pkg/db/drivers/mongo/bucketlifecycledelete.go]: Resolved in the latest commit for Delete lifecycle API

File: api/pkg/s3/bucketlifecycledelete.go (hunk 5 above)
Comment (nguptaopensds) [reply api/pkg/s3/bucketlifecycledelete.go]: Resolved in the latest commit for Delete lifecycle API

File: api/pkg/s3/bucketlifecycledelete.go (hunk 6 above)
Comment (nguptaopensds) [reply api/pkg/s3/bucketlifecycledelete.go]: Resolved in the latest commit for Delete lifecycle API

File: api/pkg/s3/bucketlifecycleget.go (hunk 12 above)
Comment (nguptaopensds) [reply api/pkg/s3/bucketlifecycleget.go]: Resolved in latest commit for GET lifecycle API

File: api/pkg/s3/bucketlifecycleget.go (hunk 13 above)
Comment (nguptaopensds) [reply api/pkg/s3/bucketlifecycleget.go]: Resolved in latest commit for GET lifecycle API

File: api/pkg/s3/bucketlifecycledelete.go [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+		for _, id := range ruleID {
+			for _, lcRule := range bucket.LifecycleConfiguration {
+				if lcRule.ID == id {
+					FoundIDArray = append(FoundIDArray, id)
+				}
+			}
+		}
+		NonFoundIDArray = differencebetweenArrays(ruleID, FoundIDArray)
+		for _, id := range NonFoundIDArray {
+			response.WriteErrorString(http.StatusBadRequest, strings.Replace("error: rule ID $1 doesn't exist \n\n", "$1", id, 1))
```
Comment (sfzeng) [inline api/pkg/s3/bucketlifecycledelete.go]: The previous id will be overlap by the next one, so only the last id of NonFoundIDArray will be left.

File: api/pkg/s3/bucketlifecycledelete.go [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+	} else {
+		response.WriteErrorString(http.StatusBadRequest, NoRuleIDForLifecycleDelete)
+		return
+	}
+	log.Log("Delete bucket lifecycle successfully.")
+}
+
+func differencebetweenArrays(rulearr []string, foundarr []string) []string {
+	var notfoundarr []string
+	for i := 0; i < 2; i++ {
```
Comment (sfzeng) [inline api/pkg/s3/bucketlifecycledelete.go]: Why need this for statement? Could you add some comments here?

File: api/pkg/s3/bucketlifecycledelete.go [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+					found = true
+					break
+				}
+			}
+			if !found {
+				notfoundarr = append(notfoundarr, ruleid)
+			}
+		}
+		if i == 0 {
+			rulearr, foundarr = foundarr, rulearr
```
Comment (sfzeng) [inline api/pkg/s3/bucketlifecycledelete.go]: same here

File: api/pkg/s3/bucketlifecycledelete.go (hunk 15 above)
Comment (nguptaopensds) [reply api/pkg/s3/bucketlifecycledelete.go]: Resolved in latest commit

File: api/pkg/s3/bucketlifecycledelete.go (hunk 16 above)
Comment (nguptaopensds) [reply api/pkg/s3/bucketlifecycledelete.go]: Resolved in latest commit

**Pull request merged**